### PR TITLE
Metrics documentation update

### PR DESCRIPTION
**Pull request by michalwski**
This PR is primarily about metrics documentation update. It also contains the following changes which may be extracted to separate PR based on review comments.

* Start the `MongoosePush.Metrics.TelemetryMetrics` child befor others to captuer events at startup
* Short metrics documentation in the code
* Add example Prometheus config for MongoosePush and docker-compose file to help with a basic setup.

### CODE REVIEW

File: lib/mongoose_push/metrics/telemetry_metrics.ex [hunk 1]
```diff
@@ -15,17 +15,28 @@ defmodule MongoosePush.Metrics.TelemetryMetrics do
         event_name: [:mongoose_push, :notification, :send],
         measurement: :time,
         buckets: [1000, 10_000, 25_000, 50_000, 100_000, 250_000, 500_000, 1000_000],
-        tags: [:status, :service, :error_category, :error_reason]
+        tags: [:status, :service, :error_category, :error_reason],
+        description:
+          "A histogram showing push notification sent times. Includes worker selection (with possible waiting if all are busy)"
```
Comment (NelsonVides) [inline lib/mongoose_push/metrics/telemetry_metrics.ex]: Great addition! 😄

File: README.md [hunk 2]
```diff
@@ -491,26 +491,78 @@ If you specify both **alert** and **data**, target device will receive both noti
...
-* **STATUS** is either `success` or `error`
-* **LENGTH** is either `100` or `250` or `500` or `1000` or `+Inf`
+* `STATUS` is `"success"` for the successful notifications or `"error"` in all other cases
+* `SERVICE` is either `"apns"` or `"fcm"`
+* `CATEGORY` is an arbitrary error category term (in case of `status="error"`) or an empty string (when `status="success"`)
+* `REASON` is an arbitrary error reason term (in case of `status="error"`) or an empty string (when `status="success"`)
+* `LE` defines the `less or equal` values for buckets, currently `1000`, `10_000`, `25_000`, `50_000`, `100_000`, `250_000`, `500_000`, `1000_000` or `+Inf`
+
+> **NOTE**
+> A mesurment of value 50_000 will be added to all buckets which are less or equal to 50_000.
```
Comment (NelsonVides) [inline README.md]:
```suggestion
> A measurement of value 50_000 will be added to all buckets which are less or equal to 50_000.
```

File: README.md [hunk 3]
```diff
@@ -491,26 +491,78 @@ If you specify both **alert** and **data**, target device will receive both noti
...
+> In this case these are buckets `50_000`, `100_000`, `250_000`, `500_000`, `1000_000` and `+Inf`
+
+This histogram metric shows the distribution of times needed to:
+1. Select a worker (this may include waiting time when all workers are busy).
+2. Send a request.
+3. Get a response from push notifications provider.
+
+##### Counters
+
+* `mongoose_push_supervisor_init_count{service=${SERVICE}}` - Counts number of push notification service supervisor starts.
```
Comment (NelsonVides) [inline README.md]:
```suggestion
* `mongoose_push_supervisor_init_count{service=${SERVICE}}` - Counts the number of push notification service supervisor starts.
```

File: README.md [hunk 4]
```diff
@@ -491,26 +491,78 @@ If you specify both **alert** and **data**, target device will receive both noti
...
+  The `SERVICE` variable can take `"apns"` or `"fcm"` as a value.
+  This metrics is update when MongoosePush starts or later when the underlying supervision tree is terminated and the error is propagate to the main application supervisor.
+* `mongoose_push_apns_state_init_count` - Counts number of APNS state initialisations.
```
Comment (NelsonVides) [inline README.md]:
```suggestion
* `mongoose_push_apns_state_init_count` - Counts the number of APNS state initialisations.
```

File: README.md [hunk 5]
```diff
@@ -491,26 +491,78 @@ If you specify both **alert** and **data**, target device will receive both noti
...
+* `mongoose_push_apns_state_terminate_count` - Counts number of APNS state terminations.
```
Comment (NelsonVides) [inline README.md]:
```suggestion
* `mongoose_push_apns_state_terminate_count` - Counts the number of APNS state terminations.
```

File: README.md [hunk 6]
```diff
@@ -491,26 +491,78 @@ If you specify both **alert** and **data**, target device will receive both noti
...
+* `mongoose_push_apns_state_get_default_topic_count` - Counts number of default topic reads from cache.
```
Comment (NelsonVides) [inline README.md]:
```suggestion
* `mongoose_push_apns_state_get_default_topic_count` - Counts the number of default topic reads from cache.
```

File: README.md [hunk 7]
```diff
@@ -491,26 +491,78 @@ If you specify both **alert** and **data**, target device will receive both noti
...
+This histogram metric shows the distribution of times needed to:
+1. Select a worker (this may include waiting time when all workers are busy).
+2. Send a request.
+3. Get a response from push notifications provider.
+
+##### Counters
+
+* `mongoose_push_supervisor_init_count{service=${SERVICE}}` - Counts number of push notification service supervisor starts.
+  The `SERVICE` variable can take `"apns"` or `"fcm"` as a value.
+  This metrics is update when MongoosePush starts or later when the underlying supervision tree is terminated and the error is propagate to the main application supervisor.
```
Comment (NelsonVides) [inline README.md]:
```suggestion
  This metrics is updated when MongoosePush starts, and later on when the underlying supervision tree is terminated and the error is propagated to the main application supervisor.
```

File: README.md [hunk 8]
```diff
@@ -491,26 +491,78 @@ If you specify both **alert** and **data**, target device will receive both noti
...
+* `mongoose_push_apns_state_get_default_topic_count` - Counts number of default topic reads from cache.
+
+#### How to quickly see all metrics
+
+```bash
+curl -k https://127.0.0.1:8443/metrics
+```
+
+The above command assumes that MongoosePush runs on `localhost` and listens on port `8443`.
+Please, mind the `HTTPS` protocol, metrics are hosted on the same port as other API.
```
Comment (NelsonVides) [inline README.md]:
```suggestion
Please, mind the `HTTPS` protocol, metrics are hosted on the same port than all the other API endpoints.
```
Or something like that, I'm not sure. I'd treat "API" as an indefinite singular noun, so "as other API" doesn't sound grammatically correct 🤔

File: README.md [hunk 9]
```diff
@@ -491,26 +491,78 @@ If you specify both **alert** and **data**, target device will receive both noti
...
+    tls_config: #The default certs used by MongoosePush are self-signed
+      insecure_skip_verify: true #For checking purposes we can ignore certs verification
+    static_configs:
+      - targets: ['mongoose-push:8443']
+        labels:
+          group: 'production'
+
+```
+
+
```
Comment (NelsonVides) [inline README.md]: Not needed empty lines
```suggestion
```

File: lib/mongoose_push/application.ex [hunk 10]
```diff
@@ -33,8 +33,10 @@ defmodule MongoosePush.Application do
     _ = check_runtime_configuration_status()
 
     # Define workers and child supervisors to be supervised
+    # The MongoosePush.Metrics.TelemetryMetrics child is started first to capture possible events
+    # when services start
     children =
-      service_children() ++ [MongoosePushWeb.Endpoint, MongoosePush.Metrics.TelemetryMetrics]
+      [MongoosePush.Metrics.TelemetryMetrics] ++ service_children() ++ [MongoosePushWeb.Endpoint]
```
Comment (NelsonVides) [inline lib/mongoose_push/application.ex:39]: Good catch 🙂

File: lib/mongoose_push/metrics/telemetry_metrics.ex [hunk 11]
```diff
@@ -15,17 +15,28 @@ defmodule MongoosePush.Metrics.TelemetryMetrics do
...
         tags: [:error_reason],
-        tag_values: fn metadata -> %{metadata | error_reason: metadata.reason} end
+        tag_values: fn metadata -> %{metadata | error_reason: metadata.reason} end,
+        description: "Counts number of APNS state terminations"
       ),
-      Telemetry.Metrics.counter("mongoose_push.apns.state.get_default_topic.count")
+      Telemetry.Metrics.counter(
+        "mongoose_push.apns.state.get_default_topic.count",
+        description: "Counts number of APNS default topic reads from the ETS cache"
+      )
```
Comment (NelsonVides) [inline lib/mongoose_push/metrics/telemetry_metrics.ex]: These are also missing the `the` articles, as in "Counts _the_ number of..."

**Review by NelsonVides — COMMENTED** [review]
Fantastic!

The only thing I'd say is that MPush docs, i.e., a single README, is getting out of control with the size, but reorganising that is on a different task 🙂 

Oh, and some articles were missing 😛

File: README.md (hunk 7 above)
Comment (kmakiela) [inline README.md]:
```suggestion
  This metric is updated when MongoosePush starts or later when the underlying supervision tree is terminated and the error is propagated to the main application supervisor.
```

File: test/docker/docker-compose.mpush.yml [hunk 12]
```diff
@@ -1,5 +1,5 @@
 # This file needs to be used along with `docker-compose.mocks.yml`:
-# docker-compose -f test/docker/docker-compose.mocks.yml -f test/docker/docker-compose.mpush.yml ...
+# PRIV=priv docker-compose -f test/docker/docker-compose.mocks.yml -f test/docker/docker-compose.mpush.yml ...
```
Comment (kmakiela) [inline test/docker/docker-compose.mpush.yml:2]: Why do we need this change?

**Review by kmakiela — COMMENTED** [review]
Great update! However, reading through 600 lines of readme is a bit much so that may be something we want to change in the near future.

File: test/docker/docker-compose.mpush.yml (hunk 12 above)
Comment (michalwski) [reply test/docker/docker-compose.mpush.yml:2]: The `docker-compose-mpush.yml` requires the `PRIV` env var. It's set for us when we do `mix text.env.up` but when running from console we need to export it manually.

**Comment by michalwski** [comment]
@NelsonVides @kmakiela thanks for the review. I 100% agree with the documentation split/refactoring. This is definitely a subject for a separate PR.

File: README.md (hunk 2 above)
Comment (jan-ciesla) [inline README.md]: Not sure if I understand this right, I think this is the other way around. The bucket contains measurements that are less or equal to the bucket size.

`More particularly they're counters which form a cumulative histogram, le stands for less than or equal to. So 26688 requests took less than or equal to 200ms, 27760 requests took less than or equal to 400ms, and there were 28860 requests in total. ` 
[https://www.robustperception.io/how-does-a-prometheus-histogram-work](https://www.robustperception.io/how-does-a-prometheus-histogram-work)

File: README.md (hunk 8 above)
Comment (jan-ciesla) [inline README.md]:
```suggestion
Please, mind the `HTTPS` protocol, metrics are hosted on the same port as other API endpoints.
```
How about this as compromise? 
`the same as` seems more grammatically correct than `the same than`

File: lib/mongoose_push/metrics/telemetry_metrics.ex (hunk 1 above)
Comment (jan-ciesla) [reply lib/mongoose_push/metrics/telemetry_metrics.ex]: `sent` or `send`? I always confuse when to use the two 😄 , but the event name has `send.time`

are they both correct or should we change one?

File: lib/mongoose_push/metrics/telemetry_metrics.ex [hunk 13]
```diff
@@ -15,17 +15,28 @@ defmodule MongoosePush.Metrics.TelemetryMetrics do
...
+      ),
       Telemetry.Metrics.counter("mongoose_push.apns.state.terminate.count",
         tags: [:error_reason],
-        tag_values: fn metadata -> %{metadata | error_reason: metadata.reason} end
+        tag_values: fn metadata -> %{metadata | error_reason: metadata.reason} end,
+        description: "Counts number of APNS state terminations"
       ),
-      Telemetry.Metrics.counter("mongoose_push.apns.state.get_default_topic.count")
+      Telemetry.Metrics.counter(
+        "mongoose_push.apns.state.get_default_topic.count",
```
Comment (jan-ciesla) [inline lib/mongoose_push/metrics/telemetry_metrics.ex]:
```suggestion
      Telemetry.Metrics.counter("mongoose_push.apns.state.get_default_topic.count",
```

just for consistency with the remaining events.

**Review by jan-ciesla — CHANGES_REQUESTED** [review]
Good stuff! I really like the PR, we just need to make sure that the part regarding `less or equal` is accurate as it can be a bit confusing.

File: lib/mongoose_push/metrics/telemetry_metrics.ex (hunk 1 above)
Comment (michalwski) [reply lib/mongoose_push/metrics/telemetry_metrics.ex]: That's probably `send` better to be consistent with the event and metric name.

File: README.md (hunk 2 above)
Comment (michalwski) [reply README.md]: Right, I messed with that. Good catch!

**Comment by michalwski** [comment]
@janciesla8818 thanks for the review. I think I applied all the comments now.

**Review by jan-ciesla — APPROVED** [review]

**Review by NelsonVides — APPROVED** [review]

**Pull request merged**